pkg/collector/corechecks/gpu/nvmlmetrics/gpm.go

```diff
@@ -0,0 +1,183 @@
+	hasPreviousSample  bool
+	currentSampleIndex int
+	tags               []string
+}
+
+func newGpmMetricsCollector(lib nvml.Interface, dev nvml.Device, tags []string) (Collector, error) {
+	var samples []nvml.GpmSample
+
+	// First, check if the device supports GPM metrics
+	support, ret := dev.GpmQueryDeviceSupport()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("failed to query GPM support for device: %s", nvml.ErrorString(ret))
+	}
+	if support.IsSupportedDevice == 0 {
+		return nil, errUnsupportedDevice
+	}
+
+	for i := 0; i < numGpmSamples; i++ {
+		sample, ret := lib.GpmSampleAlloc()
+		if ret != nvml.SUCCESS {
+			return nil, fmt.Errorf("failed to allocate GPM sample: %s", nvml.ErrorString(ret))
+		}
+		samples = append(samples, sample)
+	}
+
+	return &gpmMetricsCollector{
+		lib:     lib,
+		samples: samples,
+		device:  dev,
+		tags:    tags,
+	}, nil
+}
+
+// Name returns the name of the collector.
+func (c *gpmMetricsCollector) Name() string {
+	return gpmMetricsCollectorName
+}
+
+// Collect collects all the GPM metrics from the given NVML device.
+func (c *gpmMetricsCollector) Collect() ([]Metric, error) {
+	var err error
+
+	metricsGet, err := c.sampleAndTryGetMetrics()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get GPM metrics: %w", err)
+	}
+	if metricsGet == nil {
+		return nil, nil // Not enough samples to collect metrics, return nothing
+	}
+
+	values := make([]Metric, 0, len(allNvmlGpmMetrics))
+	for i, metric := range allNvmlGpmMetrics {
+		if metricsGet.Metrics[i].NvmlReturn != uint32(nvml.SUCCESS) {
+			err = multierror.Append(err, fmt.Errorf("failed to get metric %s: %s", metric.name, nvml.ErrorString(nvml.Return(metricsGet.Metrics[i].NvmlReturn))))
+			continue
+		}
+
+		values = append(values, Metric{
+			Name:  metric.name,
+			Value: float64(metricsGet.Metrics[i].Value),
+			Tags:  c.tags,
+		})
+	}
+
+	return values, err
+}
+
+// Close cleans up any resources used by the collector.
+func (c *gpmMetricsCollector) Close() error {
+	var err error
+	for _, sample := range c.samples {
+		sampleErr := sample.Free()
+		if sampleErr != nvml.SUCCESS {
+			err = multierror.Append(err, sampleErr)
+		}
+	}
+
+	return err
+}
+
+// currentSample returns the current GPM sample.
+func (c *gpmMetricsCollector) currentSample() nvml.GpmSample {
+	return c.samples[c.currentSampleIndex]
+}
+
+// previousSample returns the previous GPM sample.
+func (c *gpmMetricsCollector) previousSample() nvml.GpmSample {
+	return c.samples[(c.currentSampleIndex+1)%numGpmSamples]
+}
+
+// markCollectedSample marks the current sample as collected and updates the index, rotating through the samples.
+func (c *gpmMetricsCollector) markCollectedSample() {
+	c.hasPreviousSample = true
+	c.currentSampleIndex = (c.currentSampleIndex + 1) % numGpmSamples
+}
+
+// getMetrics gets a sample and, if possible, returns the GPM metrics.
+func (c *gpmMetricsCollector) sampleAndTryGetMetrics() (*nvml.GpmMetricsGetType, error) {
```

Comment:
I thought about that, but getting the sample and getting the metrics are not independent processes: we need to sample, then try get the metrics, and then mark that we have collected a sample (if we do the other two steps in the wrong order, we won't be using the correct samples). That's why I preferred to have everything in a single method, to avoid any possible bugs with that order.